src/dataflow-types/src/client/controller.rs

```diff
@@ -56,6 +60,71 @@ pub struct OrchestratorConfig {
     pub storage_addr: String,
 }
 
+fn deserialize_memory_limit<'de, D>(deserializer: D) -> Result<Option<ByteSize>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let maybe_string = Option::<String>::deserialize(deserializer)?;
+    match maybe_string {
+        None => Ok(None),
+        Some(string) => match ByteSize::from_str(&string) {
+            Ok(bs) => Ok(Some(bs)),
+            Err(_e) => {
+                use serde::de::Error;
+                Err(D::Error::invalid_value(
+                    serde::de::Unexpected::Str(&string),
+                    &"valid size in bytes",
+                ))
+            }
+        },
+    }
+}
+
+#[derive(Copy, Clone, Debug, Deserialize)]
+pub struct ClusterReplicaSizeConfig {
+    #[serde(deserialize_with = "deserialize_memory_limit")]
+    memory_limit: Option<ByteSize>,
+    cpu_limit: Option<CpuLimit>,
+    processes: NonZeroUsize,
+}
+
+#[derive(Clone, Debug)]
+pub struct ClusterReplicaSizeMap(pub HashMap<String, ClusterReplicaSizeConfig>);
+
+impl<'de> Deserialize<'de> for ClusterReplicaSizeMap {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let inner = HashMap::<_, _>::deserialize(deserializer)?;
+        Ok(Self(inner))
+    }
+}
+
+impl Default for ClusterReplicaSizeMap {
+    fn default() -> Self {
+        // {
+        //     "1": {"processes": 1},
+        //     "2": {"processes": 2},
+        //     /// ...
+        //     "16": {"processes": 16}
+        // }
+        let inner = (1..=16)
+            .map(|i| {
+                (
+                    i.to_string(),
+                    ClusterReplicaSizeConfig {
+                        memory_limit: None,
+                        cpu_limit: None,
+                        processes: NonZeroUsize::new(i).unwrap(),
+                    },
+                )
+            })
+            .collect();
+        Self(inner)
+    }
+}
```

Comment:
👍🏽

Reply:
I would have liked to make this a const, but I couldn't find a good way to construct a `HashMap` at compile time without depending on random crates.

Reply:
We could probably restrict this to powers of two, but I'm good with this as well.